coderd/workspaceupdates.go

```diff
@@ -0,0 +1,318 @@
+}
+
+// Equal does not compare agents
+func (w ownedWorkspace) Equal(other ownedWorkspace) bool {
+	return w.WorkspaceName == other.WorkspaceName &&
+		w.Status == other.Status
+}
+
+type sub struct {
+	ctx      context.Context
+	cancelFn context.CancelFunc
+
+	mu     sync.RWMutex
+	userID uuid.UUID
+	ch     chan *proto.WorkspaceUpdate
+	prev   workspacesByID
+
+	db     tailnet.UpdateQuerier
+	ps     pubsub.Pubsub
+	logger slog.Logger
+
+	psCancelFn func()
+}
+
+func (s *sub) handleEvent(ctx context.Context, event wspubsub.WorkspaceEvent, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	switch event.Kind {
+	case wspubsub.WorkspaceEventKindStateChange:
+	case wspubsub.WorkspaceEventKindAgentConnectionUpdate:
+	case wspubsub.WorkspaceEventKindAgentTimeout:
+	case wspubsub.WorkspaceEventKindAgentLifecycleUpdate:
+	default:
+		if err == nil {
+			return
+		} else {
+			// Always attempt an update if the pubsub lost connection
+			s.logger.Warn(ctx, "failed to handle workspace event", slog.Error(err))
+		}
+	}
+
+	row, err := s.db.GetWorkspacesAndAgents(ctx, s.userID)
+	if err != nil {
+		s.logger.Warn(ctx, "failed to get workspaces and agents by owner ID", slog.Error(err))
+		return
+	}
+	latest := convertRows(row)
+
+	out, updated := produceUpdate(s.prev, latest)
+	if !updated {
+		return
+	}
+
+	s.prev = latest
+	select {
+	case <-s.ctx.Done():
+		return
+	case s.ch <- out:
+	}
+}
+
+func (s *sub) start(ctx context.Context) (err error) {
+	rows, err := s.db.GetWorkspacesAndAgents(ctx, s.userID)
+	if err != nil {
+		return xerrors.Errorf("get workspaces and agents by owner ID: %w", err)
+	}
+
+	latest := convertRows(rows)
+	initUpdate, _ := produceUpdate(workspacesByID{}, latest)
+	s.ch <- initUpdate
+	s.prev = latest
+
+	cancel, err := s.ps.SubscribeWithErr(wspubsub.WorkspaceEventChannel(s.userID), wspubsub.HandleWorkspaceEvent(s.handleEvent))
+	if err != nil {
+		return xerrors.Errorf("subscribe to workspace event channel: %w", err)
+	}
+
+	s.psCancelFn = cancel
+	return nil
+}
+
+func (s *sub) Close() error {
+	s.cancelFn()
+
+	if s.psCancelFn != nil {
+		s.psCancelFn()
+	}
+
+	close(s.ch)
+	return nil
+}
+
+func (s *sub) Updates() <-chan *proto.WorkspaceUpdate {
+	return s.ch
+}
+
+var _ tailnet.Subscription = (*sub)(nil)
+
+type updatesProvider struct {
+	ps     pubsub.Pubsub
+	logger slog.Logger
+
+	ctx      context.Context
+	cancelFn func()
+}
+
+var _ tailnet.WorkspaceUpdatesProvider = (*updatesProvider)(nil)
+
+func NewUpdatesProvider(logger slog.Logger, ps pubsub.Pubsub) (tailnet.WorkspaceUpdatesProvider, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	out := &updatesProvider{
+		ctx:      ctx,
+		cancelFn: cancel,
+		ps:       ps,
+		logger:   logger,
+	}
+	return out, nil
+}
+
+func (u *updatesProvider) Close() error {
+	u.cancelFn()
+	return nil
+}
+
+func (u *updatesProvider) Subscribe(ctx context.Context, userID uuid.UUID, db tailnet.UpdateQuerier) (tailnet.Subscription, error) {
```

Comment:
Why does the `Subscribe` caller need to give the updates provider the database reference?  Can't we pass the db ref when creating the updates provider, like we do with the pubsub?

Reply:
My thought process was that this would be easier to mock, as is done in `workspaceupdates_test.go`. I didn't want to have an API that required a `database.Store` instance. To make mocking easier with that design that we'd need a second interface like `UpdatesStore` that just had the minimum query functions we need, and then that would need to get passed to `UpdatesQuerier`.

Reply:
I think that's a reasonable approach, so I'll give that a try now

Reply:
Okay, the issue is that you then somehow need to plumb either `database.Store` or some subset of it into the `UpdatesQuerier`, you can't pass it as an argument because it'd include a `database.*` type into the UpdatesQuerier interface in `tailnet`, and then from within `coderd` we only have the interface, and not the underlying type. I don't think we should be type switching on that interface just to insert the db into it.